### PR TITLE
add Time-dependent DFT

### LINKED ISSUE #300: Add `ModelMethod` description for Time-Dependent Density Functional Theory (TDDFT)
While the current schema already captures ground-state DFT in a structured way, TDDFT currently lacks a dedicated ModelMethod representation.

### PR DESCRIPTION
**Pull request by EBB2675**
## Purpose
I would like to add a general description for Time dependent Density functional theory.

## Scope

**Included** (_What is intentionally part of this PR?_)

- TDDFT model method description, that extends `ExcitedStateMethodology`, and reuses existing components such as `XCFunctional` and `Photon`

**Out of scope** (_What is explicitly not included (to avoid scope creep)?_)

- I haven't added any tests, as this is a highly small and straightforward PR. 

## Status

_Select one_

- [ ] Draft / In progress (not ready for full review)
- [x] Ready for review
- [ ] Blocked (explain below)

## Dependencies / Blockers

_List anything that may block progress or review_

- [x] None.
- [ ] Related PRs:
- [ ] Open design questions:
- [ ] External dependencies (people, decisions, data, releases):

## Testing / Validation

_How was this change validated?_

- [ ] Unit tests
- [ ] Integration tests
- [x] Manual testing (brief description):
- [ ] Not applicable (explain why):

- I haven't added any tests, as this is a highly small and straightforward PR. 

## Linked Issues / Context

- Fixes #
- Related to #
- Follow-up issues (if known):

## Reviewer Notes

We (me and Denis) aim to test this implementation soon rigorously in multiple parsers/flavours. 
There might of course be minor modifications needed in the future, in the light of our experience in parser dev.

### CODE REVIEW

File: src/nomad_simulations/schema_packages/model_method.py [hunk 1]
```diff
@@ -1345,6 +1345,149 @@ class BSE(ExcitedStateMethodology):
...
+        type=MEnum('length', 'velocity'),
+        description='Electromagnetic gauge used for coupling to the field (mostly real-time).',
+    )
+
+    def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+        super().normalize(archive, logger)
+
+        self.name = 'TDDFT'
+
+        # Default to linear_response if not specified; infer real_time from solver hint
```
Comment (JFRudzinski) [inline src/nomad_simulations/schema_packages/model_method.py]: I would not default to filling in something that isn't parsed...I now wonder if there are other instances of this?

Comment (EBB2675) [reply src/nomad_simulations/schema_packages/model_method.py]: there are indeed

Comment (EBB2675) [reply src/nomad_simulations/schema_packages/model_method.py]: removed from current PR :+1:

File: src/nomad_simulations/schema_packages/model_method.py [hunk 2]
```diff
@@ -1345,6 +1345,144 @@ class BSE(ExcitedStateMethodology):
...
+            'ALDA',
+            'AGGA',
+            'hybrid',
+            'range_separated',
+            'frequency_dependent',
+            'bootstrap',
+            'long_range_corrected',
+        ),
+        description="""
+        This field describes the effective XC response behaviour assumed in the TDDFT calculation,
```
Comment (kubanmar) [inline src/nomad_simulations/schema_packages/model_method.py]: I would maybe avoid using the words "This field" or similar, for consistency

File: src/nomad_simulations/schema_packages/model_method.py [hunk 3]
```diff
@@ -1345,6 +1345,144 @@ class BSE(ExcitedStateMethodology):
...
+        type=XCFunctional,
+        description="""
+        Reference to an `XCFunctional` describing the TDDFT kernel.
+        Typically points to the ground-state DFT XC section if the adiabatic
+        kernel matches, or to a separate XCFunctional created for the TD kernel
+        when it differs (e.g., LRC/hybrid kernel on top of semi-local ground state).
+        """,
+    )
+
+    field_ref = Quantity(
```
Comment (kubanmar) [inline src/nomad_simulations/schema_packages/model_method.py]: "external_field_ref" ?

File: src/nomad_simulations/schema_packages/model_method.py [hunk 4]
```diff
@@ -1345,6 +1345,144 @@ class BSE(ExcitedStateMethodology):
...
+        type=Photon,
+        description='External field / polarization used to drive the response or propagation.',
+    )
+
+    target = Quantity(
```
Comment (kubanmar) [inline src/nomad_simulations/schema_packages/model_method.py]: "target_quantity" or "target_property" ?

File: src/nomad_simulations/schema_packages/model_method.py [hunk 5]
```diff
@@ -1345,6 +1345,144 @@ class BSE(ExcitedStateMethodology):
...
+        # Derive kernel type from xc_kernel_ref when hybrids are clear
+        if self.kernel is None and self.xc_kernel_ref is not None:
+            try:
+                if (
+                    self.xc_kernel_ref.global_exact_exchange is not None
+                    and float(self.xc_kernel_ref.global_exact_exchange) > 0
+                ):
+                    self.kernel = 'hybrid'
+            except Exception:
+                logger.warning('Could not derive TDDFT.kernel from xc_kernel_ref.')
```
Comment (kubanmar) [inline src/nomad_simulations/schema_packages/model_method.py]: This looks like if `xc_kernel_ref.global_exact_exchange` is smaller or equal 0, `kernel` will not be set. In that case we just don"t set it, or should we log a warning?

**Review by kubanmar — APPROVED** [review]
Looks good to me, I added some minor comments. We will see if quantities are missing when we apply it to the parsers.

File: src/nomad_simulations/schema_packages/model_method.py [hunk 6]
```diff
@@ -1345,6 +1345,144 @@ class BSE(ExcitedStateMethodology):
...
+          • real_time       — explicit time propagation under a perturbation.
+        """,
+    )
+
+    kernel = Quantity(
+        type=MEnum(
+            'ALDA',
+            'AGGA',
+            'hybrid',
+            'range_separated',
```
Comment (ndaelman-hu) [inline src/nomad_simulations/schema_packages/model_method.py]: Isn't this covered by the XC functional, already?

File: src/nomad_simulations/schema_packages/model_method.py [hunk 7]
```diff
@@ -1345,6 +1345,144 @@ class BSE(ExcitedStateMethodology):
...
+          • range_separated : long/short-range split hybrids
+          • frequency_dependent : non-adiabatic kernel
+          • bootstrap / long_range_corrected : common nano/solid-state kernels
+        """,
+    )
+
+    is_tamm_dancoff = Quantity(
+        type=bool,
+        description='If True, Tamm-Dancoff approximation (TDA) is used (linear-response only).',
+    )
```
Comment (ndaelman-hu) [inline src/nomad_simulations/schema_packages/model_method.py]: If this quantity is contingent on linear-response, why not make it part of `type`?
This would save you on sanity-checking logic. If it's for searching, you could make the categories "linear-response TDA" and "linear-response other", or something along those lines.

File: src/nomad_simulations/schema_packages/model_method.py (hunk 5 above)
Comment (EBB2675) [reply src/nomad_simulations/schema_packages/model_method.py]: if `xc_kernel_ref.global_exact_exchange` is equal to 0, there are multiple options for this non hybrid `kernel`, that is simulation dependent, so not setting it here sounded appropriate

warning or info fits better :+1: will do

File: src/nomad_simulations/schema_packages/model_method.py [hunk 8]
```diff
@@ -1345,6 +1345,144 @@ class BSE(ExcitedStateMethodology):
...
+            'range_separated',
+            'frequency_dependent',
+            'bootstrap',
+            'long_range_corrected',
+        ),
+        description="""
+        This field describes the effective XC response behaviour assumed in the TDDFT calculation, 
+        whether implemented explicitly (linear-response) or implicitly via the XC potential (real-time).
+          • ALDA / AGGA     : adiabatic local or semi-local
+          • hybrid          : hybrid adiabatic kernel
```
Comment (JosePizarro3) [inline src/nomad_simulations/schema_packages/model_method.py]: maybe nice to add some references? They can be to code documentation, papers, whatever you think is appropiate to understand the choice

File: src/nomad_simulations/schema_packages/model_method.py [hunk 9]
```diff
@@ -1345,6 +1345,144 @@ class BSE(ExcitedStateMethodology):
...
+        description='If True, Tamm-Dancoff approximation (TDA) is used (linear-response only).',
+    )
+
+    solver = Quantity(
+        type=MEnum(
+            'Casida', 'Sternheimer', 'Liouville-Lanczos', 'propagation', 'unavailable'
+        ),
+        description="""
+        Numerical driver:
+          • Casida             : eigenvalue problem in particle-hole space
```
Comment (JosePizarro3) [inline src/nomad_simulations/schema_packages/model_method.py]: same comment as with the `kernel`

File: src/nomad_simulations/schema_packages/model_method.py (hunk 5 above)
Comment (EBB2675) [reply src/nomad_simulations/schema_packages/model_method.py]: or maybe i will just avoid setting the kernel at the schema level alltogether for now

File: src/nomad_simulations/schema_packages/model_method.py [hunk 10]
```diff
@@ -1345,6 +1345,144 @@ class BSE(ExcitedStateMethodology):
...
+          • Sternheimer        : frequency-domain Sternheimer
+          • Liouville-Lanczos  : iterative Liouville-Lanczos
+          • propagation        : explicit time evolution (real-time)
```
Comment (JosePizarro3) [inline src/nomad_simulations/schema_packages/model_method.py]: is `propagation` the only choice if `type="real_time"` and any of the others are for linear response?

File: src/nomad_simulations/schema_packages/model_method.py (hunk 3 above)
Comment (JosePizarro3) [reply src/nomad_simulations/schema_packages/model_method.py]: or `field_polarization_ref`

File: src/nomad_simulations/schema_packages/model_method.py (hunk 4 above)
Comment (JosePizarro3) [reply src/nomad_simulations/schema_packages/model_method.py]: voting for `target_property` or `target_response`

File: src/nomad_simulations/schema_packages/model_method.py [hunk 11]
```diff
@@ -1345,6 +1345,144 @@ class BSE(ExcitedStateMethodology):
...
+        type=Photon,
+        description='External field / polarization used to drive the response or propagation.',
+    )
+
+    target = Quantity(
+        type=MEnum('absorption', 'emission', 'EELS', 'Raman', 'nonlinear'),
+        description='Intended spectral/response target of the TDDFT input.',
+    )
+
+    time_step = Quantity(
```
Comment (JosePizarro3) [inline src/nomad_simulations/schema_packages/model_method.py]: is this not a variable of the corresponding PhysicalProperty?

File: src/nomad_simulations/schema_packages/model_method.py [hunk 12]
```diff
@@ -1345,6 +1345,144 @@ class BSE(ExcitedStateMethodology):
...
+            'ETRS',
+            'Magnus',
+            'Runge-Kutta',
+            'Lanczos',
+            'split_operator',
+        ),
+        description='Time integrator used for real-time propagation.',
+    )
+
+    gauge = Quantity(
```
Comment (JosePizarro3) [inline src/nomad_simulations/schema_packages/model_method.py]: mm what do you mean here?

File: src/nomad_simulations/schema_packages/model_method.py [hunk 13]
```diff
@@ -1345,6 +1345,144 @@ class BSE(ExcitedStateMethodology):
...
+        description='Intended spectral/response target of the TDDFT input.',
+    )
+
+    time_step = Quantity(
+        type=np.float64,
+        unit='second',
+        description='Time step used for real-time propagation.',
+    )
+
+    n_steps = Quantity(
```
Comment (JosePizarro3) [inline src/nomad_simulations/schema_packages/model_method.py]: same comment as for `time_step`

File: src/nomad_simulations/schema_packages/model_method.py [hunk 14]
```diff
@@ -1345,6 +1345,144 @@ class BSE(ExcitedStateMethodology):
...
+        type=np.int32,
+        description='Number of time steps in real-time propagation.',
+    )
+
+    propagator = Quantity(
```
Comment (JosePizarro3) [inline src/nomad_simulations/schema_packages/model_method.py]: same comment as for `time_step`

this is also related with numerical_settings.Mesh (cc @ndaelman-hu)

**Review by JosePizarro3 — APPROVED** [review]
All good, just minor stuff

File: src/nomad_simulations/schema_packages/model_method.py (hunk 8 above)
Comment (EBB2675) [reply src/nomad_simulations/schema_packages/model_method.py]: will do :+1:

File: src/nomad_simulations/schema_packages/model_method.py [hunk 15]
```diff
@@ -1345,6 +1345,135 @@ class BSE(ExcitedStateMethodology):
...
+
+    xc_kernel_ref = Quantity(
+        type=XCFunctional,
+        description="""
+        Reference to an `XCFunctional` describing the TDDFT kernel.
+        Typically points to the ground-state DFT XC section if the adiabatic
+        kernel matches, or to a separate XCFunctional created for the TD kernel
+        when it differs (e.g., LRC/hybrid kernel on top of semi-local ground state).
+        """,
+    )
```
Comment (ndaelman-hu) [inline src/nomad_simulations/schema_packages/model_method.py]: So, the DFT section has to be populated together with this one, correct? Maybe add this to the section's pydocs. (First time reading the description, I thought you were using the reference to describe a workflow, so the extra clarification would be nice.)
I would then also reference to `DFT` instead. Other corrections like Hubbard, Vandermonde, VdW corrections can all still be applied on top.

File: src/nomad_simulations/schema_packages/model_method.py (hunk 11 above)
Comment (ndaelman-hu) [reply src/nomad_simulations/schema_packages/model_method.py]: I think it fits well here. I don't think the objective is to capture the propagation (that should be added to the workflow level) during the real-time simulation, but just to define a simulation parameter.

File: src/nomad_simulations/schema_packages/model_method.py (hunk 8 above)
Comment (ndaelman-hu) [reply src/nomad_simulations/schema_packages/model_method.py]: there are references in the section-level docs

File: src/nomad_simulations/schema_packages/model_method.py [hunk 16]
```diff
@@ -1345,6 +1345,135 @@ class BSE(ExcitedStateMethodology):
...
+        if self.type == 'real_time' and (
+            self.time_step is None or self.n_steps is None
+        ):
+            logger.warning(
+                'TDDFT real_time mode without time_step or n_steps may be incomplete.'
+            )
+        if self.type == 'linear_response' and self.is_tamm_dancoff is None:
+            logger.warning(
+                'TDDFT linear_response set but is_tamm_dancoff not specified.'
+            )
```
Comment (ndaelman-hu) [inline src/nomad_simulations/schema_packages/model_method.py]: You aren't covering the other case, where `is_tamm_dancoff` is set, but `type == real-time`.
If the edge cases become sufficiently complex, I would in fact add a test. You can then move this logic out to a dedicated function.

**Review by ndaelman-hu — APPROVED** [review]
Slight restructuring suggestions + few comments to other reviewers

File: src/nomad_simulations/schema_packages/model_method.py (hunk 13 above)
Comment (ndaelman-hu) [reply src/nomad_simulations/schema_packages/model_method.py]: Previous comment accidentally got deleted:

> I think it fits well here. I don't think the objective is to capture the propagation (that should be added to the workflow level) during the real-time simulation, but just to define a simulation parameter.

File: src/nomad_simulations/schema_packages/model_method.py (hunk 14 above)
Comment (ndaelman-hu) [reply src/nomad_simulations/schema_packages/model_method.py]: Yes, I agree that semantically it has more of a numerical settings nature.
@EBB2675 if you add it under a `NumericalSettings` child section, consider adding a reference from here to there.

Comment (EBB2675) [reply src/nomad_simulations/schema_packages/model_method.py]: will do :+1:

**Review by TLCFEM — APPROVED** [review]

**Comment by EBB2675** [comment]
- Removed kernel-specific fields from TDDFT (notably `kernel` and `xc_kernel_ref`) to keep TDDFT aligned with the current GW/BSE-style method sections, and to defer XC-kernel details to a **future** workflow-level PR.

- Kept only the essential "flavour" switch via `type = {'linear_response', 'real_time'}` to distinguish frequency-domain response from time-propagation TDDFT.

- Introduced `approximation` = {'full', 'TDA'}, and dropped linear-response-only flags like `is_tamm_dancoff`. In the future, we can enrich this Enum with more approximations.

- Removed real-time "required settings" checks (`time_step`, `n_steps`) since those belong in `numerical_settings.py` and are explicitly postponed in this PR.

- Added light consistency warnings to prevent impossible combinations (e.g., type='real_time' with solver='Casida', or approximation='TDA' with type='real_time').

At the end, it turned into a light method class similar to already existing excited state methodology sections, numerical knobs will come slowly in future PRs :+1: 

Please take a last look :rose: 
@JFRudzinski @ndaelman-hu @JosePizarro3 @kubanmar

**Review by JFRudzinski — APPROVED** [review]
